File: day19/notDuplicate.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <algorithm>
#include <random>
#include <ctime>

int singleNumber(const std::vector<int>& nums) {
    int result = 0;
    for (int num : nums) {
        result ^= num; // XOR cancels out pairs
    }
    return result;
}

int main() {
    std::mt19937 rng(static_cast<unsigned int>(std::time(nullptr)));
    std::uniform_int_distribution<int> lengthDist(10, 20);
    std::uniform_int_distribution<int> valueDist(0, 99);

    int length = lengthDist(rng);

    // Use unordered_set to ensure unique values
    std::unordered_set<int> uniqueVals;
    while (uniqueVals.size() < static_cast<size_t>(length)) {
        uniqueVals.insert(valueDist(rng));
    }

    std::vector<int> vals(uniqueVals.begin(), uniqueVals.end());
    int single = vals.back();
    vals.pop_back(); // Remove the one that appears once

    std::vector<int> nums;
    for (int val : vals) {
        nums.push_back(val);
        nums.push_back(val);
    }
    nums.push_back(single); // Add the single element
    std::shuffle(nums.begin(), nums.end(), rng);

    std::cout << "Input: ";
    for (int n : nums) std::cout << n << " ";
    std::cout << "\nExpected single number: " << single << "\n";
    std::cout << "Computed single number: " << singleNumber(nums) << "\n";

    return 0;
}

File: day19/not_duplicate.rs
use rand::{seq::SliceRandom, thread_rng, Rng};
use std::collections::HashSet;

fn single_number(nums: &[i32]) -> i32 {
    nums.iter().fold(0, |acc, &x| acc ^ x) // XOR cancels out duplicates
}

fn main() {
    let mut rng = thread_rng();
    let length = rng.gen_range(10..=20);

    // HashSet ensures unique values
    let mut unique_vals = HashSet::new();
    while unique_vals.len() < length {
        unique_vals.insert(rng.gen_range(0..100));
    }

    let mut vals: Vec<i32> = unique_vals.into_iter().collect();
    let single = vals.pop().unwrap(); // One value appears once

    let mut nums: Vec<i32> = vals
        .iter()
        .flat_map(|&x| vec![x, x]) // Each remaining value appears twice
        .collect();

    nums.push(single);
    nums.shuffle(&mut rng);

    println!("Input: {:?}", nums);
    println!("Expected single number: {}", single);
    println!("Computed single number: {}", single_number(&nums));
}
